module/x/peggy/abci.go

```diff
@@ -130,50 +130,92 @@ func ValsetSlashing(ctx sdk.Context, k keeper.Keeper, params types.Params) {
 
-			// Don't slash validators who joined after valset is created
+		// SLASH BONDED VALIDTORS who didn't attest valset request
+		currentBondedSet := k.StakingKeeper.GetBondedValidatorsByPower(ctx)
+		for _, val := range currentBondedSet {
 			consAddr, _ := val.GetConsAddr()
 			valSigningInfo, exist := k.SlashingKeeper.GetValidatorSigningInfo(ctx, consAddr)
-			if exist && valSigningInfo.StartHeight > int64(vs.Nonce) {
-				continue
-			}
 
-			// Don't slash validators who are unbonded and UNBOND_SLASHING_WINDOW has passed
-			if val.UnbondingHeight > 0 && vs.Nonce > uint64(val.UnbondingHeight)+params.UnbondSlashingValsetsWindow {
-				continue
-			}
+			//  Slash validator ONLY if he joined after valset is created
+			if exist && valSigningInfo.StartHeight < int64(vs.Nonce) {
+				// Check if validator has confirmed valset or not
+				found := false
+				for _, conf := range confirms {
+					if conf.EthAddress == k.GetEthAddress(ctx, val.GetOperator()) {
+						found = true
+						break
+					}
+				}
+				// slash validators for not confirming valsets
+				if !found {
+					cons, _ := val.GetConsAddr()
+					k.StakingKeeper.Slash(ctx, cons, ctx.BlockHeight(), val.ConsensusPower(), params.SlashFractionValset)
+					if !val.IsJailed() {
+						k.StakingKeeper.Jail(ctx, cons)
+					}
 
-			// Check if validator has confirmed valset or not
-			found := false
-			for _, conf := range confirms {
-				if conf.EthAddress == k.GetEthAddress(ctx, val.GetOperator()) {
-					found = true
-					break
 				}
 			}
+		}
 
-			// slash validators for not confirming valsets
-			if !found {
-				cons, _ := val.GetConsAddr()
-				k.StakingKeeper.Slash(ctx, cons, ctx.BlockHeight(), val.ConsensusPower(), params.SlashFractionValset)
-				k.StakingKeeper.Jail(ctx, cons)
+		// SLASH UNBONDING VALIDATORS who didn't attest valset request
+		blockTime := ctx.BlockTime().Add(k.StakingKeeper.GetParams(ctx).UnbondingTime)
+		blockHeight := ctx.BlockHeight()
+		unbondingValIterator := k.StakingKeeper.ValidatorQueueIterator(ctx, blockTime, blockHeight)
+		defer unbondingValIterator.Close()
+
+		// All unbonding validators
+		for ; unbondingValIterator.Valid(); unbondingValIterator.Next() {
+			unbondingValidators := k.GetUnbondingvalidators(unbondingValIterator.Value())
+
+			for _, valAddr := range unbondingValidators.Addresses {
+				addr, err := sdk.ValAddressFromBech32(valAddr)
+				if err != nil {
+					panic(err)
+				}
+				validator, exist := k.StakingKeeper.GetValidator(ctx, sdk.ValAddress(addr))
+				valConsAddr, _ := validator.GetConsAddr()
+				valSigningInfo, exist := k.SlashingKeeper.GetValidatorSigningInfo(ctx, valConsAddr)
 
+				// Only slash validators who joined after valset is created and they are unbonding and UNBOND_SLASHING_WINDOW didn't passed
+				if exist && valSigningInfo.StartHeight < int64(vs.Nonce) && validator.IsUnbonding() && vs.Nonce < uint64(validator.UnbondingHeight)+params.UnbondSlashingValsetsWindow {
```

Comment:
valset nonces being generated as block heights isn't the best idea, we should probably add a block height parameter and not count on the nonces.

Reply:
I think we should leave this for another pr though.